### ISSUE TITLE
rest-client-reactive URL-encodes PathParam - non-reactive dependency doesn't

### ISSUE THREAD

**Issue by higgs01** [quarkusio/quarkus#38123]
### Describe the bug

I have a RestClient like this:
```java
@RegisterRestClient(configKey = "test-restclient")
@Consumes(MediaType.APPLICATION_JSON)
public interface TestRestClient {

    @POST
    @Path("/{test}")
    String runFunc(@HeaderParam("Authorization") String authorization, @PathParam("test") String test);

}
```

If I use `quarkus-rest-client-jackson` everything works fine. If i use `quarkus-rest-client-reactive-jackson` the PathParam is suddendly URL-Encoded (and therefore points to a invalid resource beause `/` is encoded).

### Expected behavior

`quarkus-rest-client-reactive-jackson` and `quarkus-rest-client-jackson` behave the same and don't escape the path-param.

### Actual behavior

With `quarkus-rest-client-reactive-jackson` the URL-Param `test/asdf` is URL-Encoded and therefore rendered as `test%2Fasdf`. Since I'm trying to call the endpoint `test/asdf` the request now returns a 404 because the path is wrong. With `quarkus-rest-client-jackson` everything works fine.

### How to Reproduce?

See code from above. If neccesairy i can also provide a reproduction-project.

### Output of `java -version`

Java version: 17.0.9, vendor: Eclipse Adoptium

### Quarkus version or git rev

3.6.4

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.5 (57804ffe001d7215b5e7bcb531cf83df38f93546)

### Additional information

Additional Dependencies:
- io.quarkus:quarkus-arc

**Comment by geoand** [comment]
> If neccesairy i can also provide a reproduction-project.

That would be lovely!

**Comment by higgs01** [comment]
@geoand I've uploaded a POC to https://github.com/higgs01/quarkus-reprod-38123 - if you have any questions feel free to reach out.

**Comment by geoand** [comment]
Thanks for the sample!

I think we've actually discussed this before and come to the conclusion that what the Reactive REST does here is the proper behavior. 

cc @FroMage @cescoffier

**Comment by cescoffier** [comment]
Yes, it's tricky, but it's the proper behavior, as the path parameter should represent a single path segment (at least, that's my understanding).

If we do not encode it, path parameters using spaces or other characters needing encoding will not work.
A path parameter containing `/` also somewhat abuses the path parameter, as it should be a single segment of the URL. 

Except if you have very dynamic URLs where you do not know the number of segments, the proper approach is to have a single path parameter per segment.

**Comment by higgs01** [comment]
Thank you for your response 👍🏻 I guess it was just the inconsistency between the reactive and non-reactive REST-Client that confused me. 

> Except if you have very dynamic URLs where you do not know the number of segments, the proper approach is to have a single path parameter per segment.

What would be the approach if you have highly dynamic urls?

**Comment by geoand** [comment]
> What would be the approach if you have highly dynamic urls?

Is this something you need?

**Comment by higgs01** [comment]
No, I was just curious. For me this issue is resolved and can be closed.

**Comment by geoand** [comment]
Gotcha, thanks.

**Comment by sascha-friederich** [comment]
We ran into the same issue - and have 'highly dynamic urls' - is there another possibility apart from 
 QuarkusRestClientBuilder.newBuilder().baseUri(URI.create('highly dynamic urls')) ??

**Comment by osenbg** [comment]
We have the same Problem with 'highly dynamic urls'

**Comment by geoand** [comment]
Do you have an example application that shows that in action?

**Comment by FroMage** [comment]
The Jakarta REST spec mentions that this is the proper behaviour, and any normal path parameter represents a single segment. If you want them to accept more than one segment you can use regexes: https://jakarta.ee/specifications/restful-ws/4.0/jakarta-restful-ws-spec-4.0#uritemplates

> ```java
> @Path("widgets/{path:.+}")
> public class Widget {
>     ...
> }
> ```
> 
> In the above example the Widget resource class will be matched for any request whose path starts with widgets and contains at least one more path segment; the value of the path parameter will be the request path following widgets. E.g. given the request path `widgets/small/a` the value of `path` would be `small/a`.

I'm not entirely sure whether the REST client will encode path segments when targeting a path parameter which is a regex. The client spec doesn't say: https://download.eclipse.org/microprofile/microprofile-rest-client-4.0/microprofile-rest-client-spec-4.0.html

Worth trying and see what happens. And since the spec doesn't say, we could take a stance and implement it so that it does not do any `/` encoding, if that's what people feel it should do. It would be coherent with the server's spec.

**Comment by davidreis97** [comment]
Not entirely sure why this was closed @geoand, this is a valid use-case even if the person who you're talking to at the moment doesn't immediately need it... It's also not feasible to ask for the code as we're building enterprise apps that are closed-source, it's not like this is a strange issue that is hard to reproduce

**Comment by geoand** [comment]
> It's also not feasible to ask for the code as we're building enterprise apps that are closed-source

I don't see anyone asking for the exact application.

> it's not like this is a strange issue that is hard to reproduce

If that is the case, by all means add a sample and we reexamine the issue.

>  this is a valid use-case 

True as that may be, it doesn't mean the declarative REST Client can be all things for all people. That's why I asked for a sample so we can see things in action instead of on a theoretical basis.

**Comment by davidreis97** [comment]
@geoand It's all good, our use-case is handled by the `@Encoded` annotation, just trying to make the point that just because someone doesn't have an immediate use-case it's not necessarily a bad idea 🙂

**Comment by geoand** [comment]
> just because someone doesn't have an immediate use-case and code example it's not necessarily a bad idea 

Absolutely! And for the record, no one said it was :)
